engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/UnresolvedConstructor.java

```diff
@@ -137,17 +173,39 @@ Object instantiateCached(
         UnresolvedConstructor unresolved,
         @Cached("expectedType") Type cachedType,
         @Cached("unresolved.asPrototype()") UnresolvedConstructor prototype,
-        @Cached("expectedType.getConstructors().get(prototype.getName())") AtomConstructor c,
-        @Cached("build(prototype.descs,EXEC_MODE,ARGS_MODE)") InvokeFunctionNode invoke) {
+        @Cached("expectedType.getConstructors().get(prototype.getName())") AtomConstructor c) {
       if (c == null) {
         return null;
       } else {
+        if (invoke == null) {
+          CompilerDirectives.transferToInterpreterAndInvalidate();
+          Lock lock = getLock();
+          lock.lock();
+          try {
+            invoke = insert(buildInvokeNode(prototype));
+            where = prototype.where.getSourceSection();
+            if (invoke.getId() != null) {
+              notifyInsertedWithCheck();
+            }
+          } finally {
+            lock.unlock();
+          }
+        }
         var fn = c.getConstructorFunction();
         var r = invoke.execute(fn, frame, state, unresolved.args);
         return r;
       }
     }
 
+    private void notifyInsertedWithCheck() {
```

Comment:
The here-in provided changes are not 100% correct. They seem to deliver some instrumentation events, but probably there are corner cases when they don't work well. Anyway, there are some events generated, so hopefully @4e6 can investigate what to do with `ProgramExecutionSupport` even at current state.